Add sample for separate image samplers

SaschaWillems opened the pull request:
## Description

This PR adds a basic api sample demonstrating how to separate images and samplers. This is similar to the LunarG sample from the archived repository.

The sample uses one image and two different samplers, one with linear and one with nearest filtering, which can be selected at runtime. 

It also comes with a small tutorial/readme describing the relevant parts of the application and shader code and a small comparison with a combined setup.

The sample was tested on Windows 10 with an NVIDIA RTX 2060. It's based on the texture loading sample, so it should work any pretty much every platform.

Fixes #302

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

SaschaWillems committed:
Set theme jekyll-theme-slate

SaschaWillems committed:
Merge remote-tracking branch 'origin/master'

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master'

SaschaWillems committed:
Merge branch 'KhronosGroup:master' into master

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master'

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master'

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master'

SaschaWillems committed:
New sample for separate image and samplers

SaschaWillems committed:
Made relevant code parts more explicit

SaschaWillems committed:
Code comments

SaschaWillems committed:
Added short tutorial/readme

SaschaWillems committed:
Code formatting in readme

SaschaWillems committed:
Reworded some paragraphs in the readme.

SaschaWillems committed:
Added separate image sampler sample to sample listing readme

SaschaWillems committed:
Cleaned up shaders

SaschaWillems committed:
Clang format

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master'

SaschaWillems committed:
Merge branch 'master' into separate_image_sampler

SaschaWillems committed:
Added shader files to project structure via CMake

gpx1000 commented on samples/api/separate_image_sampler/separate_image_sampler.cpp:
```diff
@@ -0,0 +1,500 @@
+
+#include "separate_image_sampler.h"
+
+SeparateImageSampler::SeparateImageSampler()
```
missing a few member variables that aren't initialized.

gpx1000 commented on samples/api/separate_image_sampler/separate_image_sampler.cpp:
```diff
@@ -0,0 +1,500 @@
+
+	vertex_buffer.reset();
+	index_buffer.reset();
+	uniform_buffer_vs.reset();
```
these are unique_ptr no need to explicitly call reset they will go away when out of scope.

gpx1000 commented on samples/api/separate_image_sampler/separate_image_sampler.h:
```diff
@@ -0,0 +1,86 @@
+	VkDescriptorSetLayout sampler_descriptor_set_layout;
+
+	SeparateImageSampler();
+	~SeparateImageSampler();
```
need override (or at least virtual) on destructor.

Also either virtual or override on the below, no need for both; prefer override.

SaschaWillems replied:
Are you sure on this? To my knowledge there's no need to specify virtual or override on the derived class' destructor. But I'm not 100% sure.

gpx1000 replied:
https://github.com/isocpp/CppCoreGuidelines/issues/721#issuecomment-598878968

While it's not a requirement it's good practice.  The above link is a good description of the reasoning.
I'd add my own to say when you do a lot of unit testing being explicit really helps.  In general you don't really know how your class will be used.  For the cost of a no-op to allow for future devs to be able to derive from your class, delete your class and get their derived class to call destructor correctly, etc.
The only real place it matters is in the base class, where any virtual function MUST have a virtual destructor.  The override being used for any function that overrides the virtual base class is correct however, making an exception for the destructor is a source of confusion for a lot of devs.  That's why the guidelines were changed.  For me personally, it's just simpler to just develop to the best practices of always adding the proper notation even when it's a no-op.  Code remains clean and correct and never have problems with confusion.

SaschaWillems replied:
Thank you very much for the detailed rundown. That makes a lot of sense.

gpx1000 commented on samples/api/separate_image_sampler/separate_image_sampler.cpp:
```diff
@@ -0,0 +1,500 @@
+		VkViewport viewport = vkb::initializers::viewport((float) width, (float) height, 0.0f, 1.0f);
+		vkCmdSetViewport(draw_cmd_buffers[i], 0, 1, &viewport);
+
+		VkRect2D scissor = vkb::initializers::rect2D(width, height, 0, 0);
```
static_cast<int>() on width/height.

SaschaWillems replied:
Thanks for pointing this out. Will change this. I'm wondering why Visual Studio doesn't output any warnings for me on this :/

gpx1000 commented on samples/api/separate_image_sampler/separate_image_sampler.cpp:
```diff
@@ -0,0 +1,500 @@
+	        0);
+
+	// Load shaders
+	std::array<VkPipelineShaderStageCreateInfo, 2> shader_stages;
```
init the array prior to being used.  Ideally just add initializer list with the load_shader call directly in there.

gpx1000 reviewed (dismissed):
small warning issues but otherwise looks good to me runs well here.

SaschaWillems committed:
Suggested code changes:

- Initialize member variables
- Remove explicit buffer reset
- Mark destructor as override

gary-sweet reviewed (dismissed):
Looks good to me

JoseEmilio-ARM commented on samples/api/separate_image_sampler/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+# limitations under the License.
+#
+
+# @todo
```
can remove `todo`?

SaschaWillems replied:
Good catch. Somehow missed these :)

JoseEmilio-ARM commented on samples/api/separate_image_sampler/separate_image_sampler.cpp:
```diff
@@ -0,0 +1,495 @@
+
+void SeparateImageSampler::setup_descriptor_set()
+{
+	// @todo: comment, compare to combined setup (or better do this in the readme.md)
```
can remove `todo`?

JoseEmilio-ARM commented on samples/api/separate_image_sampler/separate_image_sampler.h:
```diff
@@ -0,0 +1,86 @@
+
+	Texture texture;
+
+	// @todo: comment
```
can remove `todo`?

SaschaWillems committed:
Removed left-over todos
